go-controller/pkg/network-controller-manager/network_controller_manager.go

```diff
@@ -0,0 +1,276 @@
+		Lock:            rl,
+		LeaseDuration:   time.Duration(config.MasterHA.ElectionLeaseDuration) * time.Second,
+		RenewDeadline:   time.Duration(config.MasterHA.ElectionRenewDeadline) * time.Second,
+		RetryPeriod:     time.Duration(config.MasterHA.ElectionRetryPeriod) * time.Second,
+		ReleaseOnCancel: true,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: func(ctx context.Context) {
+				klog.Infof("Won leader election; in active mode")
+				start := time.Now()
+				defer func() {
+					end := time.Since(start)
+					metrics.MetricMasterReadyDuration.Set(end.Seconds())
+				}()
+
+				if err := cm.Init(stopChan); err != nil {
+					klog.Error(err)
+					cancel()
+					return
+				}
+
+				if err = cm.Run(ctx); err != nil {
+					klog.Error(err)
+					cancel()
+					return
+				}
+			},
+			OnStoppedLeading: func() {
+				//This node was leader and it lost the election.
+				// Whenever the node transitions from leader to follower,
+				// we need to handle the transition properly like clearing
+				// the cache.
+				klog.Infof("No longer leader; exiting")
+				cancel()
+			},
+			OnNewLeader: func(newLeaderName string) {
+				if newLeaderName != cm.identity {
+					klog.Infof("Lost the election to %s; in standby mode", newLeaderName)
+				}
+			},
+		},
+	}
+
+	leaderelection.SetProvider(ovnkubeMasterLeaderMetricsProvider{})
+	leaderElector, err := leaderelection.NewLeaderElector(lec)
+	if err != nil {
+		return err
+	}
+
+	cm.wg.Add(1)
+	go func() {
+		leaderElector.Run(ctx)
+		klog.Infof("Stopped leader election")
+		cm.wg.Done()
+	}()
+
+	return nil
+}
+
+// NewNetworkControllerManager creates a new OVN controller manager to manage all the controller for all networks
+func NewNetworkControllerManager(ovnClient *util.OVNClientset, identity string, wf *factory.WatchFactory,
+	libovsdbOvnNBClient libovsdbclient.Client, libovsdbOvnSBClient libovsdbclient.Client,
+	recorder record.EventRecorder, wg *sync.WaitGroup) *NetworkControllerManager {
+	podRecorder := metrics.NewPodRecorder()
+
+	return &NetworkControllerManager{
+		client: ovnClient.KubeClient,
+		kube: &kube.Kube{
+			KClient:              ovnClient.KubeClient,
+			EIPClient:            ovnClient.EgressIPClient,
+			EgressFirewallClient: ovnClient.EgressFirewallClient,
+			CloudNetworkClient:   ovnClient.CloudNetworkClient,
+		},
+		watchFactory: wf,
+		recorder:     recorder,
+		nbClient:     libovsdbOvnNBClient,
+		sbClient:     libovsdbOvnSBClient,
+		podRecorder:  &podRecorder,
+
+		wg:             wg,
+		identity:       identity,
+		ovnControllers: make(map[string]NetworkController),
+	}
+}
+
+// Init initializes the controller manager and create/start default controller
+func (cm *NetworkControllerManager) Init(stopChan <-chan struct{}) error {
+	err := cm.configureSCTPSupport()
+	if err != nil {
+		return err
+	}
+
+	err = cm.enableOVNLogicalDataPathGroups()
+	if err != nil {
+		return err
+	}
+
+	cm.configureMetrics(stopChan)
```

Comment:
I would probably move `configureMetrics` to the very beginning of the Init method. Previously this was in the beginning of the `StartClusterMaster` method; some of the metrics relates to the NBDB and SBDB timestamps and we are already writing to NBDB with `enableOVNLogicalDataPathGroups`. Just in case.